polars/polars-lazy/polars-plan/src/logical_plan/optimizer/stack_opt.rs

```diff
@@ -53,12 +53,19 @@ impl StackOptimizer {
                 while let Some(expr_node) = scratch.pop() {
                     exprs.push(expr_node);
                     // traverse all subexpressions and add to the stack
-                    let expr = expr_arena.get(expr_node);
+                    let expr = unsafe { expr_arena.get_unchecked(expr_node) };
                     expr.nodes(&mut exprs);
                 }
 
                 // process the expressions on the stack and apply optimizations.
                 while let Some(current_expr_node) = exprs.pop() {
+                    {
+                        let expr = unsafe { expr_arena.get_unchecked(current_expr_node) };
+                        // don't apply rules to `col`, `lit` etc.
+                        if expr.is_leaf() {
```

Comment:
This will be most expressions and it is pointless to optimize them as they are atomic.